kep/399-sysched-scoring/README.md

```diff
@@ -0,0 +1,161 @@
+
+## Design Details: Overall architecture
+
+Our plugin utilizes system call profiles of pods obtained through their security contexts in the form of seccomp profiles. Our plugin computes a score for each feasible node by comparing the system call profile of the pod with the system call profiles of the existing pods in the node. Our plugin keeps track of the placements of pods in the entire cluster. To compute the scores, the plugin uses a metric called Extraneous System Call (ExS). The ExS metic captures the system calls that a pod does not need, but is present in the node.  Our plugin then returns the computed normalized ExS scores to be combined with other scores in Kubernetes for ranking nodes. If a pod has no seccomp profile associated with it, then our plugin effectively is a noop. The following figure shows the integration of our plugin with the Kubernetes scheduler.
+
+<p align="center">
+<img src="images/arch.png" style="width:70%">
+</p>
+
+<p align="center">
+<b>Figure 3: Overall architecture</b>
+</p>
+
+### Scheduling Metrics - Extraneous System Call (ExS)
+
+At the heart of SySched is a single-valued score used to quantify the extraneous system call exposure of a pod on a node. Informally, ExS captures the additional system calls accessible on a node that the pod itself does not use. This score reflects the potential “danger” a pod must face when running on the target node as vulnerabilities in these extraneous system calls from neighboring pods can jeopardize its own security. In the following figure, we can obtain the ExS for pod C1, C2, and C3 for a node with a kernel of nine system calls.
+
+<p align="center">
+<img src="images/exs.png" style="width:60%">
+</p>
+
+<p align="center">
+<b>Figure 4: Example ExS score calculation</b>
+</p>
+
+### Scheduling Plugin
+
+We developed a new plugin to implement the scoring extension API point in the Kubernetes scheduler while leaving other filters and scoring plugins intact. We leverage existing filtering and scoring operations of the Kubernetes scheduler to handle other aspects of pod placement such as spreading the pods and ensuring resource availability.
+
+ There are two main components of our plugin: 1) a lightweight in memory state store that holds the up-to-date node to pod mapping and 2) mechanisms for retrieving the system call sets used by pods and for computing the ExS scores. The state store is implemented as a thread that listens to events of a pod’s lifecycle, e.g., creation, stop, destruction, run, etc. Using these event notifications, this thread tracks where pods are currently running in a cluster. When a pod stops running on a node, our plugin also removes it from its internal mapping. While the entire state of pod placement in a cluster can be retrieved dynamically by querying the API server, this can be an expensive task to do per pod scheduling event, hence we choose to maintain this state internally in our scheduler.
```

Comment:
These two stanzas do not reveal essential info to illustrate how a pod is scored. Some of the info like how it relates with filter and other score plugins are assumed to be known by readers, so we usually don't need to mention it.

I guess the second stanza here is trying to illustrate that there is a cache defined to syn with pod events, so that we keep an up-to-date state about what types of syscalls are present on what nodes. If it's the case, we can simplify it a bit.